[Frontend] unit tests for node server

eterna2 opened the pull request:
WIP: Adding basic unit tests for frontend node server.

PS: I noted some errors inside because seems like my new deployment cannot access the pod logs from my S3 buckets. Writing some unit tests to narrow down where are the bugs (could be my deployment issue also - cuz it worked last time, stop working after I upgrade the version).

See #2407 
cc: @Bobgy

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/kubeflow/pipelines/2745)
<!-- Reviewable:end -->

Bobgy commented on frontend/server/aws-helper-test.ts:
```diff
@@ -0,0 +1,62 @@
+// Copyright 2019 Google LLC
```
Our convention is to name like `aws-helper.test.ts`

eterna2 commented:
I need break up the big `server.ts` file so that unit test can be done more easily.

Bobgy commented:
@eterna2 Thanks, that sounds reasonable.

Note that there are some ongoing work on https://github.com/kubeflow/pipelines/pull/2690/files which also touches node server. Beware of that and I recommend don't do too much refactoring that may conflict with it.

eterna2 commented:
## Summary of changes:
- unit tests for `aws-helper` and `minio-helper` (I found my bug, aws iam credential should work even when the session token expires now)
- broke up `setup.ts` into `app.ts` (registering all the middleware), `handlers/*.ts` (defining the individual middleware), and `configs.ts` (functions to parse `argv` and env vars into a config singleton)
- e2e unit test for the express app:
  - test for `gcs`, `s3`, `minio`, and `http/https` artifacts
  - test for `healthz`
  - test for `index.html`

## Impacts on other PRs
- minor impact for #2690, but not a difficult change, just move `server.ts` changes to `handlers/*.ts`, no impact for `k8s-helper.ts` as it is not changed.

## TODOs (in future PR?)
- refractor `k8s-helper` (current state makes it impossible to do any unit test or mocking)
  - BLOCKS:
    - unit test for `workflow-helper`
    - e2e unit test for express app: for k8s-related routes (can't test, cuz can't mock k8s api)
- probably shld add a real e2e test for the proxies in the integration test

eterna2 commented:
I updated `@types/node` because `lib` to fix https://github.com/microsoft/TypeScript/issues/32333

And updated `@kubernetes/node-client` to maintain compatibility with the updated `@types/node`.

This shld fix the docker build errors.

Bobgy commented on frontend/server/handlers/healthz.ts:
```diff
@@ -0,0 +1,63 @@
+import fetch from 'node-fetch';
+import { Handler } from 'express';
+
+export interface IHealthzStats {
```
Sorry, I think I forgot to identify this problem also in last PR.
This repo's convention is to never prefix interface with `I`, just use `HealthzStats` sounds good.

eterna2 replied:
fixed. renamed all the interfaces.

eterna2 replied:
renamed all interfaces.

Bobgy commented on frontend/server/app.ts:
```diff
@@ -0,0 +1,161 @@
+
+  start(port?: number | string) {
+    port = port || this.options.server.port;
+    this.httpServer = this.app.listen(port, () => {
```
nit: also throw an error if http server is already initialized?

eterna2 replied:
fixed.

Bobgy commented on frontend/server/app.ts:
```diff
@@ -0,0 +1,161 @@
+  }
+
+  close() {
+    return this.httpServer && this.httpServer.close();
```
nit: and clear httpServer instance?

eterna2 replied:
fixed

Bobgy commented on frontend/server/app.ts:
```diff
@@ -0,0 +1,161 @@
+  app: Application;
+  httpServer?: Server;
+
+  constructor(public readonly options: IConfigs) {
```
I'm not familiar with node common patterns, is using constructor with heavy initialization logic common?
I'd personally prefer using a pure function, e.g. `initUIServer(options)` to create an instance of UIServer. Extra benefit is that, the creator function can be async too (I guess it's arguable whether we need this or not).

This is just discussion. What do you think?

eterna2 replied:
classes are just syntactic sugars in js. I think there are no right common patterns (as in it is fine either way). 

The main reason I did this is so that I write less code in my unit test. It is easier for me to do the setup and teardown if I only need to track a single object.

But I think this is perfectly fine too.
```js
const app = initUIServer(options);
const server = app.listen(port, ()=>console.info(`listening to port:${port}.`);
server.close();
```

Bobgy commented:
/test kubeflow-pipeline-sample-test

eterna2 commented on frontend/.gitignore:
```diff
@@ -24,3 +24,9 @@ backstop_data/bitmaps_test/
+coverage
+
+# vscode
+.vscode
```
fixed

eterna2 commented:
- Renamed all the interface (removed `I`)
- Rebased to include the recent PR with tensorboard versioning
- Added comments and fixed some lint/format errors

Bobgy commented:
kubeflow-pipeline-sample-test test is flaky, we can retry later

Bobgy commented on frontend/server/app.test.ts:
```diff
@@ -0,0 +1,333 @@
+    app && app.close();
+  });
+
+  it('api server is not ready', done => {
```
@eterna2 thanks for these great efforts in testing! They help a lot!

I have to add some style guide here. Recommended test message structure is the following:
* describe messages describe what subject or conditions you are testing.
* you can use describe nesting to divide tests by groups, e.g.
```
describe('UIServer', () => {
  describe('index.html', () => {
    it('responds', ...);
    it('responds with kubeflow flag and client injected when in kubeflow deployment', ...);
  });
});
```
* it messages describe expected behavior after certain actions. Usually it messages should read like an english sentence like `it('reports api server is not ready', () => {...})`.
* it message should convey roughly about expected behavior too. `it` refers to the subject you are testing.

The guideline is consistent with existing UI unit tests in this repo and google internal guides (unfortunately they are not public), but I think this is fairly common practice. I don't have a good resource on this, you can grasp an idea of that by reading this: https://kentcdodds.com/blog/write-fewer-longer-tests (read his test messages, your tests are structured at a good size).

Naming things is hard... which we all know. So if you don't feel comfortable changing this, I can follow up after this is merged too. What do you think?

eterna2 replied:
sure. I can make the changes.

Bobgy replied:
Thank you! That would be awesome!

Bobgy commented:
@eterna2 Here's how to integrate these tests in our CI.

This is the entrypoint for kubeflow-pipeline-frontend-test:
https://github.com/kubeflow/pipelines/blob/488805c68371bb83213c2444dde10d444d6b0bcc/frontend/package.json#L60

You can add an entry for calling server unit tests. It's fine doing it without coverage report. I can add coverage data later.

eterna2 commented on frontend/server/utils.ts:
```diff
@@ -13,6 +13,25 @@
+  if (!!port) {
+    return `${schema}://${host}${namespace}:${port}`;
+  }
+  return `${schema}://${host}${namespace}`;
```
agree. Will make this change.

eterna2 commented on frontend/server/handlers/tslint.json:
```diff
@@ -0,0 +1,7 @@
+{
```
actually I am not too sure. But at least in vscode, I need it to be inside the folder for the correct rules to be applied. But I can be wrong. Let me try.

eterna2 commented on frontend/server/handlers/tensorboard.ts:
```diff
@@ -0,0 +1,101 @@
+  } catch (err) {
+    res.status(500).send('Failed to list Tensorboard pods: ' + JSON.stringify(err));
+  }
+};
```
roger that.

eterna2 commented on frontend/server/handlers/index-html.ts:
```diff
@@ -0,0 +1,74 @@
+    }
+    return content;
+  } catch (err) {
+    console.error(`Failed to load index.html: ${err}`);
```
👌

eterna2 commented on frontend/server/handlers/pod-logs.ts:
```diff
@@ -0,0 +1,69 @@
+  if (argoOptions.archiveLogs) {
+    const minioClientOptions =
+      argoOptions.archiveArtifactory === 'minio' ? artifactsOptions.minio : artifactsOptions.aws;
+    podLogsHandler.setFallbackHandler(
```
yeah. Let me relook at this again.

eterna2 commented on frontend/server/configs.ts:
```diff
@@ -0,0 +1,214 @@
+}
+
+/** converts string to bool */
+const asBool = (value: string) => ['true', '1'].indexOf(value.toLowerCase()) >= 0;
```
oh yeah. I am still in the es2015 era. haha.

eterna2 commented on frontend/server/app.ts:
```diff
@@ -0,0 +1,199 @@
+   * @param port optionally overwrite the provided port to listen to.
+   */
+  start(port?: number | string) {
+    if (!!this.httpServer) {
```
just to force the value to bool, just to be neat. But not really necessary, unless u are returning a predicate result (which is suppose to be bool) for a function.

```js
typeof !!"abc" === 'boolean'
typeof "abc" === 'string'
```

Bobgy replied:
Got it, unless returning a predicate result. Shall we omit unnecessary operators?

eterna2 commented:
Resolved all ur comments. Rewrote pod logs handler to be more functional. And added corresponding unit tests.

Also updated test:prow.

Bobgy commented:
Awesome, I will review today

Bobgy commented:
/hold
It seems e2e tests are failing. let me retry first
kubeflow-pipeline-e2e-test

Bobgy commented:
/test kubeflow-pipeline-e2e-test

eterna2 commented on frontend/server/minio-helper.test.ts:
```diff
@@ -11,69 +11,103 @@
-    (awsInstanceProfileCredentials.ok as jest.Mock).mockClear();
+    (MinioClient as any).mockReset();
+    (awsInstanceProfileCredentials.getCredentials as jest.Mock).mockReset();
+    (awsInstanceProfileCredentials.ok as jest.Mock).mockReset();
```
ok fixed all the tests.

Bobgy commented:
/hold cancel
Thanks! Looks great

Bobgy commented:
/lgtm
